File: kotlin.web.demo.server/static/components/highlighting/HighlightingProvider.js
/*
 * Copyright 2000-2012 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Created with IntelliJ IDEA.
 * User: Natalia.Ukhorskaya
 * Date: 3/29/12
 * Time: 1:56 PM
 */

var HighlighterDecorator = (function () {

    function HighlighterDecorator() {
        var currentHighlighter = null;

        var instance = {
            getHighlighting:function (configurationType, programText, callback) {
                if (currentHighlighter != null) {
                    currentHighlighter.getHighlighting(configurationType, programText, callback);
                }
            },
            onHighlight:function (data) {

            },
            setHighlighter:function (highlighter) {
                if (currentHighlighter != null) {
                    currentHighlighter.onHighlight = null;
                    currentHighlighter.onFail = null;
                }
                currentHighlighter = highlighter;
                currentHighlighter.onHighlight = function (data, callback) {
                    instance.onHighlight(data);
                    callback(data);
                };
                currentHighlighter.onFail = function (exception) {
                    instance.onFail(exception);
                };
            },
            onFail:function (message) {
            }
        };

        return instance;
    }

    return HighlighterDecorator;
})();


/*var arrayOfHighlighters = [];
 arrayOfHighlighters.push(Configuration.mode.CLIENT.highlighter);
 arrayOfHighlighters.push(Configuration.mode.SERVER.highlighter);

 var i = 0;
 for (i = 0; i < arrayOfHighlighters.length; ++i) {
 arrayOfHighlighters[i].onHighlight = function (data, callback) {
 instance.onHighlight(data);
 callback(data);
 };
 arrayOfHighlighters[i].onFail = function (exception) {
 instance.onFail(exception);
 };
 }*/

File: kotlin.web.demo.frontend/src/providers/HighlightingProvider.kt
/*
 * Copyright 2000-2015 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package providers

import model.File
import model.Project
import views.editor.Error
import utils.Object

class HighlightingProvider(
        private val onSuccess: (dynamic) -> Unit,
        private val onFail: (String, String) -> Unit
) {

    fun getHighlighting(project: Project, callback: (Map<File, Array<Error>>) -> Unit, finallyCallback: (() -> Unit)?) {
        ajax(
                //runConf is unused parameter. It's added to url for useful access logs
                url = generateAjaxUrl("highlight", hashMapOf("runConf" to project.confType)),
                success = { data ->
                    try {
                        val errors = getErrorsMapFromObject(data, project)
                        onSuccess(errors);
                        callback(errors);
                    } catch (e: Throwable) {
                        console.log(e);
                    }
                },
                dataType = DataType.JSON,
        type = HTMLRequestType.POST,
        data = json( "project" to JSON.stringify(project) ),
        timeout = 10000,
        error = { jqXHR, textStatus, errorThrown ->
            try {
                if (jqXHR.responseText != null && jqXHR.responseText != "") {
                    onFail(jqXHR.responseText, "");
                } else {
                    onFail(textStatus + " : " + errorThrown, "");
                }
            } catch (e: Throwable) {
                console.log(e)
            }
        },
        complete = {
            finallyCallback?.invoke()
        }
        )
    }
}